torchao/quantization/prototype/qat.py

```diff
@@ -0,0 +1,286 @@
+        self.qparams_precision: torch.dtype = qparams_precision
+
+    def prepare(
+        self,
+        model: torch.nn.Module,
+        *args: Any,
+        **kwargs: Any
+    ) -> torch.nn.Module:
+        replace_linear_8da4w_qat(
+            model,
+            self.groupsize,
+            self.padding_allowed,
+            self.precision,
+            self.precision,
+        )
+        return model
+
+    def convert(
+        self,
+        model: torch.nn.Module,
+        *args: Any,
+        **kwargs: Any
+    ) -> torch.nn.Module:
+        # TODO: replace Int8DynActInt4WeightQATLinear -> Int8DynActInt4WeightLinear
+        pass
+
+
+class Int8DynActInt4WeightQATLinear(torch.nn.Linear):
+    """
+    This module implements a linear layer with int8 dynamic per token fake
+    quantized activations with int4 fake quantized grouped per channel weights.
+
+    args:
+        groupsize: the number of elements in each quantized group for weights
+        weight_precision: precision of weights
+        qparams_precision: precision of per group scales and zero points
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = False,
+        groupsize: int = 256,
+        weight_precision: torch.dtype = torch.float32,
+        qparams_precision: torch.dtype = torch.float32,
+    ) -> None:
+        super().__init__(
+            in_features,
+            out_features,
+            bias,
+            device=None,
+            dtype=weight_precision,
+        )
+        assert (
+            in_features % groupsize == 0
+        ), f"require in_features:{in_features} % groupsize:{groupsize} == 0"
+        assert not bias, "require bias=False"
+        self.groupsize = groupsize
+        self.qparams_precision = qparams_precision
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # activations: int8 dynamic asymmetric quant
+        (act_qmin, act_qmax) = self._get_qmin_qmax(8)
+        (act_scales, act_zp) = _choose_qparams_per_token_asymmetric(
+            x, torch.int8,  # dtype not used
+        )
+        x_fq = fake_quantize_per_token(
+            x, act_scales, act_zp, act_qmin, act_qmax,
+        )
+
+        # weights: int4 grouped per channel symmetric quant
+        (weight_qmin, weight_qmax) = self._get_qmin_qmax(4)
+        (weight_scales, weight_zp) = get_group_qparams_symmetric(
+            self.weight, 4, self.groupsize, self.qparams_precision,
+        )
+        w_fq = fake_quantize_per_channel_group(
+            self.weight,
+            weight_scales,
+            weight_zp,
+            weight_qmin,
+            weight_qmax,
+            self.groupsize,
+        )
+        return torch.nn.functional.linear(x_fq, w_fq)
+
+    def _get_qmin_qmax(self, n_bit: int):
+        qmin = -(2 ** (n_bit - 1))
+        qmax = 2 ** (n_bit - 1) - 1
+        return (qmin, qmax)
+
+
+def replace_linear_8da4w_qat(
```

Comment:
this seems to be the same as replace_linear_8da4w, maybe we want to abstract out a helper function, to be less error prone